core/templates/pages/exploration-player-page/learner-experience/ratings-and-recommendations.component.spec.ts

```diff
@@ -0,0 +1,160 @@
+import { RatingsAndRecommendationsComponent } from './ratings-and-recommendations.component';
+
+describe('Ratings and recommendations component', () => {
+  let fixture: ComponentFixture<RatingsAndRecommendationsComponent>;
+  let componentInstance: RatingsAndRecommendationsComponent;
+  let alertsService: AlertsService;
+  let learnerViewRatingService: LearnerViewRatingService;
+  let urlService: UrlService;
+  let userService: UserService;
+
+  const mockNgbPopover = jasmine.createSpyObj(
+    'NgbPopover', ['close', 'toggle']);
+
+
+  class MockWindowRef {
+    nativeWindow = {
+      location: {
+        search: '',
+        pathname: '/path/name',
+        reload: () => {}
+      }
+    };
+  }
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [
+        HttpClientTestingModule,
+        NgbPopoverModule
+      ],
+      declarations: [
+        RatingsAndRecommendationsComponent,
+        MockLimitToPipe
+      ],
+      providers: [
+        AlertsService,
+        LearnerViewRatingService,
+        UrlService,
+        UserService,
+        {
+          provide: WindowRef,
+          useClass: MockWindowRef
+        }
+      ],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(RatingsAndRecommendationsComponent);
+    componentInstance = fixture.componentInstance;
+    alertsService = TestBed.inject(AlertsService);
+    learnerViewRatingService = TestBed.inject(LearnerViewRatingService);
+    urlService = TestBed.inject(UrlService);
+    userService = TestBed.inject(UserService);
+  });
+
+  it('should initialize', fakeAsync(() => {
+    let collectionId = 'collection_id';
+    let userRating = 5;
+    let mockOnRatingUpdated = new EventEmitter<void>();
+
+    spyOn(urlService, 'getCollectionIdFromExplorationUrl').and.returnValue(
+      collectionId);
+    spyOn(learnerViewRatingService, 'getUserRating').and.returnValue(
+      userRating);
+    spyOn(alertsService, 'addSuccessMessage');
+    spyOn(learnerViewRatingService, 'init').and.callFake(
+      (callb: (rating: number) => void) => {
+        callb(userRating);
+      });
+    spyOnProperty(learnerViewRatingService, 'onRatingUpdated').and.returnValue(
+      mockOnRatingUpdated);
+
+    componentInstance.questionPlayerConfig = null;
+
+    componentInstance.ngOnInit();
+    mockOnRatingUpdated.emit();
+    tick();
+
+    expect(componentInstance.userRating).toEqual(userRating);
+    expect(alertsService.addSuccessMessage).toHaveBeenCalled();
+    expect(learnerViewRatingService.getUserRating).toHaveBeenCalled();
+    expect(componentInstance.collectionId).toEqual(collectionId);
+  }));
+
+  it('should toggle popover when user clicks on rating stars', () => {
+    componentInstance.feedbackPopOver = mockNgbPopover;
+
+    componentInstance.togglePopover();
+
+    expect(componentInstance.feedbackPopOver.toggle).toHaveBeenCalled();
```

Comment:
Here and below, is there any way to check the popover's final "state" rather than use toHaveBeenCalled()? The latter is weaker.

Reply:
No, because we are using a mock for ngbPopover.

Reply:
OK. I suggest leaving a comment to explain this here (and elsewhere where toHaveBeenCalled() is used).

Reply:
Done

Reply:
Ah -- sorry I was unclear. I think one more thing you'd want to explain (in these comments) is why we need to use a mock for the popover in the first place, since that seems to be a choice on our part.

Reply:
Hi @ashutoshc8101 -- just flagging this comment since it hasn't been addressed yet. Thanks!

Reply:
Done